File: ChessApp/ChessEngine/Board/PieceTable.cs
using System.Collections.Generic;
using System.Linq;

namespace ChessEngine.Board
{
    internal sealed class PieceTable
    {
        private readonly LinkedListNode<PieceOnBoard>[] _table = new LinkedListNode<PieceOnBoard>[Board.SquareCount];

        private readonly LinkedList<PieceOnBoard>[] _playerPieces = new LinkedList<PieceOnBoard>[2];

        public PieceTable(List<PieceOnBoard> pieces)
        {
            foreach (var player in new Player[] { Player.Black, Player.White })
            {
                var piecesLinkedList = CreatePlayerPiecesLinkedList(pieces, player);
                _playerPieces[(int)player] = piecesLinkedList;
                FillTableWithPieces(piecesLinkedList);
            }
        }

        public bool IsEmpty(int squareIndex) => _table[squareIndex] == null;

        public IEnumerable<PieceOnBoard> GetPieces(Player player) => _playerPieces[(int)player];

        public BoardSquare GetKingSquare(Player player) => _playerPieces[(int)player].First.Value.square;

        public PieceOnBoard GetPieceAt(int squareIndex) => _table[squareIndex].Value;

        public void MovePiece(int originIndex, int destIndex)
        {
            var nodeToMove = _table[originIndex];
            var p = nodeToMove.Value;
            nodeToMove.Value = new PieceOnBoard(p.player, p.piece, new BoardSquare(destIndex));
            _table[destIndex] = nodeToMove;
            _table[originIndex] = null;
        }

        public void RemovePieceAt(int squareIndex)
        {
            var nodeToRemove = _table[squareIndex];
            nodeToRemove.List.Remove(nodeToRemove);
            _table[squareIndex] = null;
        }

        public void InsertPiece(PieceOnBoard pieceOnBoard)
        {
            var linkedList = _playerPieces[(int)pieceOnBoard.player];
            var squareIndex = pieceOnBoard.square.IntValue;
            if (pieceOnBoard.piece == Piece.King)
            {
                _table[squareIndex] = linkedList.AddFirst(pieceOnBoard);
            }
            else
            {
                _table[squareIndex] = linkedList.AddLast(pieceOnBoard);
            }
        }

        private void FillTableWithPieces(LinkedList<PieceOnBoard> pieces)
        {
            var currentNode = pieces.First;
            while (currentNode != null)
            {
                _table[currentNode.Value.square.IntValue] = currentNode;
                currentNode = currentNode.Next;
            }
        }

        private static LinkedList<PieceOnBoard> CreatePlayerPiecesLinkedList(
            List<PieceOnBoard> pieces,
            Player player)
        {
            var kingData = pieces
                .Find(x => x.piece == Piece.King && x.player == player);

            var piecesLinkedList = new LinkedList<PieceOnBoard>(
                pieces.Where(x => x.piece != Piece.King && x.player == player));
            piecesLinkedList.AddFirst(kingData);
            return piecesLinkedList;
        }
    }
}

File: android/core/src/main/java/ru/maksimbulva/chess/core/engine/board/PieceTable.kt
package ru.maksimbulva.chess.core.engine.board

import ru.maksimbulva.chess.core.PlayerMap
import ru.maksimbulva.chess.core.collections.BufferedLinkedList
import ru.maksimbulva.chess.core.engine.Piece
import ru.maksimbulva.chess.core.engine.Player

private typealias LinkedListNode = BufferedLinkedList.Node<PieceOnBoard>

internal class PieceTable(pieces: Iterable<PieceOnBoard>) {

    private val table = arrayOfNulls<LinkedListNode>(size = Board.BOARD_CELL_COUNT)

    private val playerPieces = PlayerMap(
        blackPlayerValue = createPlayerPiecesLinkedList(pieces, Player.Black),
        whitePlayerValue = createPlayerPiecesLinkedList(pieces, Player.White)
    )

    init {
        fillTableWithPieces(playerPieces.get(Player.Black))
        fillTableWithPieces(playerPieces.get(Player.White))
    }

    fun isEmpty(cell: Cell) = table[cell.index] == null

    fun isNotEmpty(cell: Cell) = table[cell.index] != null

    fun pieces(player: Player): Iterable<PieceOnBoard> = playerPieces.get(player)

    fun kingCell(player: Player) = playerPieces.get(player).head?.value?.cell

    fun pieceAt(cell: Cell): PieceOnBoard? = table[cell.index]?.value

    fun movePiece(fromCell: Cell, toCell: Cell) {
        val nodeToMove = pieceNodeAt(fromCell)
        val pieceToMove = nodeToMove.value
        nodeToMove.value = PieceOnBoard(pieceToMove.player, pieceToMove.piece, toCell)
        table[fromCell.index] = null
        table[toCell.index] = nodeToMove
    }

    fun removePieceAt(cell: Cell) {
        val nodeToRemove = pieceNodeAt(cell)
        playerPieces.get(nodeToRemove.value.player).remove(nodeToRemove)
        table[cell.index] = null
    }

    fun insertPiece(pieceOnBoard: PieceOnBoard) {
        val piecesLinkedList = playerPieces.get(pieceOnBoard.player)
        table[pieceOnBoard.cell.index] = if (pieceOnBoard.piece == Piece.King) {
            piecesLinkedList.insertFirst(pieceOnBoard)
        } else {
            piecesLinkedList.insertAfterHead(pieceOnBoard)
        }
    }

    fun updatePiece(cell: Cell, newPiece: Piece) {
        with (pieceNodeAt(cell)) {
            table[cell.index]!!.value = PieceOnBoard(value.player, newPiece, value.cell)
        }
    }

    private fun fillTableWithPieces(piecesLinkedList: BufferedLinkedList<PieceOnBoard>) {
        var currentNode = piecesLinkedList.head
        while (currentNode != null) {
            table[currentNode.value.cell.index] = currentNode
            currentNode = currentNode.next
        }
    }

    private fun pieceNodeAt(cell: Cell) = table[cell.index]!!
}

private fun createPlayerPiecesLinkedList(
    pieces: Iterable<PieceOnBoard>,
    player: Player
): BufferedLinkedList<PieceOnBoard> {
    val king = pieces.first { it.piece == Piece.King && it.player == player }
    return BufferedLinkedList<PieceOnBoard>().apply {
        insertFirst(king)
        pieces.asSequence()
            .filter { it.piece != Piece.King && it.player == player }
            .forEach(this::insertAfterHead)
    }
}
